### PR TITLE
create product category

### PR DESCRIPTION
**Pull request by DevWael**
I have changed the functionality of the importer class to be able to add product categories with it's nested categories #118

### CODE REVIEW

File: includes/class-dropshipping-woocommerce-importer.php [hunk 1]
```diff
@@ -6,30 +6,29 @@
...
 
-if (class_exists('WC_Product_Importer', false)) :
+if ( class_exists( 'WC_Product_Importer', false ) ) :
 
 	/**
 	 * Knawat_Dropshipping_Woocommerce_Importer Class.
 	 */
-	class Knawat_Dropshipping_Woocommerce_Importer extends WC_Product_Importer
-	{
+	class Knawat_Dropshipping_Woocommerce_Importer extends WC_Product_Importer {
```
Comment (HazemKhaled) [inline includes/class-dropshipping-woocommerce-importer.php:31]: Codacy found an issue: [The class Knawat_Dropshipping_Woocommerce_Importer has an overall complexity of 162 which is very high. The configured complexity threshold is 50.](https://app.codacy.com/gh/Knawat/dropshipping-woocommerce/pullRequest?prid=6189509)

File: includes/class-dropshipping-woocommerce-importer.php [hunk 2]
```diff
@@ -243,67 +248,65 @@ public function import()
...
+				$new_product['description'] = isset( $product->description->$default_lang ) ? sanitize_textarea_field( $product->description->$default_lang ) : '';
 
-				if ($active_plugins['qtranslate-x'] && !empty($active_langs)) {
+				if ( $active_plugins['qtranslate-x'] && ! empty( $active_langs ) ) {
 
-					$new_product['name'] = '';
+					$new_product['name']        = '';
 					$new_product['description'] = '';
-					$categories = array();
+					$categories                 = array();
```
Comment (HazemKhaled) [inline includes/class-dropshipping-woocommerce-importer.php:296]: Codacy found an issue: [Avoid unused local variables such as '$categories'.](https://app.codacy.com/gh/Knawat/dropshipping-woocommerce/pullRequest?prid=6189509)

File: includes/class-dropshipping-woocommerce-importer.php [hunk 3]
```diff
@@ -469,138 +513,139 @@ public function get_formatted_product($product)
...
 		 * @param WC_Product $variation Product instance.
-		 * @param array      $data    Item data.
+		 * @param array $data Item data.
+		 *
 		 * @return WC_Product|WP_Error
 		 * @throws Exception If data cannot be set.
 		 */
-		protected function set_variation_data(&$variation, $data)
-		{
+		protected function set_variation_data( &$variation, $data ) {
```
Comment (HazemKhaled) [inline includes/class-dropshipping-woocommerce-importer.php:543]: Codacy found an issue: [The method set_variation_data() has a Cyclomatic Complexity of 14. The configured cyclomatic complexity threshold is 10.](https://app.codacy.com/gh/Knawat/dropshipping-woocommerce/pullRequest?prid=6189509)

File: includes/class-dropshipping-woocommerce-importer.php [hunk 4]
```diff
@@ -609,129 +654,133 @@ public function get_import_params()
...
-		 * @since 2.0.2
 		 * @param array $lang_object Object of values with lang keys
+		 *
 		 * @return string $formated_value Formated string with language wrappers.
+		 * @since 2.0.2
 		 */
-		function attribute_languagfy($lang_object)
-		{
-			if (empty($lang_object)) {
+		function attribute_languagfy( $lang_object ) {
```
Comment (HazemKhaled) [inline includes/class-dropshipping-woocommerce-importer.php:747]: Codacy found an issue: [The method attribute_languagfy() has a Cyclomatic Complexity of 15. The configured cyclomatic complexity threshold is 10.](https://app.codacy.com/gh/Knawat/dropshipping-woocommerce/pullRequest?prid=6189509)

File: includes/class-dropshipping-woocommerce-importer.php [hunk 5]
```diff
@@ -312,94 +315,136 @@ public function get_formatted_product($product)
...
 						}
 					}
 				}
 			}
 			$new_product['tags'] = array();
-			if ($active_plugins['qtranslate-x'] && !empty($active_langs)){
-				foreach($product->categories as $category){
-					foreach($active_langs as $active_lang){
-						if(isset($category->name->$active_lang)){
+			$z                   = 1;
```
Comment (HazemKhaled) [inline includes/class-dropshipping-woocommerce-importer.php:364]: Codacy found an issue: [Avoid unused local variables such as '$z'.](https://app.codacy.com/gh/Knawat/dropshipping-woocommerce/pullRequest?prid=6189509)

File: includes/class-dropshipping-woocommerce-importer.php [hunk 6]
```diff
@@ -609,129 +654,133 @@ public function get_import_params()
...
 		 * @param array $product_id Product ID for get attribute options
 		 * @param array $attribute_name Attribute name for get attribute options
+		 *
 		 * @return array $terms Attribute options
+		 * @since 2.0.2
 		 */
-		function get_existing_attribute_values($product_id, $attribute_name)
-		{
-			if (empty($product_id) || empty($attribute_name)) {
+		function get_existing_attribute_values( $product_id, $attribute_name ) {
```
Comment (HazemKhaled) [inline includes/class-dropshipping-woocommerce-importer.php:701]: Codacy found an issue: [The method get_existing_attribute_values() has a Cyclomatic Complexity of 17. The configured cyclomatic complexity threshold is 10.](https://app.codacy.com/gh/Knawat/dropshipping-woocommerce/pullRequest?prid=6189509)

File: includes/class-dropshipping-woocommerce-importer.php [hunk 7]
```diff
@@ -6,30 +6,29 @@
...
 	exit;
 }
 
 /**
  * Include dependencies.
  */
-if (!class_exists('WC_Product_Importer', false)) {
-	include_once dirname(WC_PLUGIN_FILE) . '/includes/import/abstract-wc-product-importer.php';
+if ( ! class_exists( 'WC_Product_Importer', false ) ) {
+	include_once dirname( WC_PLUGIN_FILE ) . '/includes/import/abstract-wc-product-importer.php';
```
Comment (HazemKhaled) [inline includes/class-dropshipping-woocommerce-importer.php:23]: Codacy found an issue: [The use of function dirname() is discouraged](https://app.codacy.com/gh/Knawat/dropshipping-woocommerce/pullRequest?prid=6189509)

File: includes/class-dropshipping-woocommerce-importer.php [hunk 8]
```diff
@@ -99,102 +96,110 @@ public function import()
...
-				return array('status' => 'fail', 'message' => $this->data->get_error_message());
-			} else if (!isset($this->data->products) && !('single' === $this->import_type && isset($this->data->product))) {
-				knawat_dropshipwc_logger('[GET_PRODUCTS_FROM_API_ERROR] ' . $api_url . ' ' . print_r($this->data, true));
-				return array('status' => 'fail', 'message' => __('Something went wrong during get data from Knawat MP API. Please try again later.', 'dropshipping-woocommerce'));
+			if ( is_wp_error( $this->data ) ) {
+				knawat_dropshipwc_logger( '[GET_PRODUCTS_FROM_API_ERROR] ' . $api_url . ' ' . $this->data->get_error_message() );
+
+				return array( 'status' => 'fail', 'message' => $this->data->get_error_message() );
+			} else if ( ! isset( $this->data->products ) && ! ( 'single' === $this->import_type && isset( $this->data->product ) ) ) {
+				knawat_dropshipwc_logger( '[GET_PRODUCTS_FROM_API_ERROR] ' . $api_url . ' ' . print_r( $this->data, true ) );
```
Comment (HazemKhaled) [inline includes/class-dropshipping-woocommerce-importer.php:126]: Codacy found an issue: [The use of function print_r() is discouraged](https://app.codacy.com/gh/Knawat/dropshipping-woocommerce/pullRequest?prid=6189509)

File: includes/class-dropshipping-woocommerce-importer.php [hunk 9]
```diff
@@ -65,32 +64,30 @@ class Knawat_Dropshipping_Woocommerce_Importer extends WC_Product_Importer
...
 			$this->import_type = $import_type;
-			$this->params = wp_parse_args($params, $default_args);
+			$this->params      = wp_parse_args( $params, $default_args );
 
 			$this->mp_api = new Knawat_Dropshipping_Woocommerce_API();
 		}
 
-		public function import()
-		{
+		public function import() {
```
Comment (HazemKhaled) [inline includes/class-dropshipping-woocommerce-importer.php:90]: Codacy found an issue: [The method import() has an NPath complexity of 1140864. The configured NPath complexity threshold is 200.](https://app.codacy.com/gh/Knawat/dropshipping-woocommerce/pullRequest?prid=6189509)

File: includes/class-dropshipping-woocommerce-importer.php [hunk 10]
```diff
@@ -204,34 +209,34 @@ public function import()
...
 						$data['updated'][] = $result['id'];
 					} else {
 						$data['imported'][] = $result['id'];
 					}
 					$product_id = $result['id'];
-					if (!empty($variations)) {
+					if ( ! empty( $variations ) ) {
 
-						foreach ($variations as $vindex => $variation) {
+						foreach ( $variations as $vindex => $variation ) {
```
Comment (HazemKhaled) [inline includes/class-dropshipping-woocommerce-importer.php:224]: Codacy found an issue: [Avoid unused local variables such as '$vindex'.](https://app.codacy.com/gh/Knawat/dropshipping-woocommerce/pullRequest?prid=6189509)

File: includes/class-dropshipping-woocommerce-importer.php (hunk 9 above)
Comment (HazemKhaled) [inline includes/class-dropshipping-woocommerce-importer.php:90]: Codacy found an issue: [The method import() has 161 lines of code. Current threshold is set to 100. Avoid really long methods.](https://app.codacy.com/gh/Knawat/dropshipping-woocommerce/pullRequest?prid=6189509)

File: includes/class-dropshipping-woocommerce-importer.php [hunk 11]
```diff
@@ -243,67 +248,65 @@ public function import()
...
 			return $data;
-			
-		}
 
-		public function get_formatted_product($product)
-		{
+		}
 
-			if (empty($product)) {
+		public function get_formatted_product( $product ) {
```
Comment (HazemKhaled) [inline includes/class-dropshipping-woocommerce-importer.php:258]: Codacy found an issue: [The method get_formatted_product() has a Cyclomatic Complexity of 62. The configured cyclomatic complexity threshold is 10.](https://app.codacy.com/gh/Knawat/dropshipping-woocommerce/pullRequest?prid=6189509)

Comment (HazemKhaled) [inline includes/class-dropshipping-woocommerce-importer.php:258]: Codacy found an issue: [The method get_formatted_product() has an NPath complexity of 16949578200. The configured NPath complexity threshold is 200.](https://app.codacy.com/gh/Knawat/dropshipping-woocommerce/pullRequest?prid=6189509)

File: includes/class-dropshipping-woocommerce-importer.php (hunk 5 above)
Comment (HazemKhaled) [inline includes/class-dropshipping-woocommerce-importer.php:364]: Codacy found an issue: [Avoid variables with short names like $z. Configured minimum length is 3.](https://app.codacy.com/gh/Knawat/dropshipping-woocommerce/pullRequest?prid=6189509)

File: includes/class-dropshipping-woocommerce-importer.php [hunk 12]
```diff
@@ -312,94 +315,136 @@ public function get_formatted_product($product)
...
 					}
 				}
 			}
 			$new_product['tags'] = array();
-			if ($active_plugins['qtranslate-x'] && !empty($active_langs)){
-				foreach($product->categories as $category){
-					foreach($active_langs as $active_lang){
-						if(isset($category->name->$active_lang)){
+			$z                   = 1;
+			$cat                 = $new_terms = array();
```
Comment (HazemKhaled) [inline includes/class-dropshipping-woocommerce-importer.php:365]: Codacy found an issue: [Avoid unused local variables such as '$cat'.](https://app.codacy.com/gh/Knawat/dropshipping-woocommerce/pullRequest?prid=6189509)

File: includes/class-dropshipping-woocommerce-importer.php (hunk 7 above)
Comment (HazemKhaled) [inline includes/class-dropshipping-woocommerce-importer.php:23]: Codacy found an issue: ["include_once" statement detected. File manipulations are discouraged. Concatenating is forbidden.](https://app.codacy.com/gh/Knawat/dropshipping-woocommerce/pullRequest?prid=6189509)

File: includes/class-dropshipping-woocommerce-importer.php (hunk 4 above)
Comment (HazemKhaled) [inline includes/class-dropshipping-woocommerce-importer.php:747]: Codacy found an issue: [The method attribute_languagfy() has an NPath complexity of 1440. The configured NPath complexity threshold is 200.](https://app.codacy.com/gh/Knawat/dropshipping-woocommerce/pullRequest?prid=6189509)

File: includes/class-dropshipping-woocommerce-importer.php (hunk 9 above)
Comment (HazemKhaled) [inline includes/class-dropshipping-woocommerce-importer.php:90]: Codacy found an issue: [The method import() has a Cyclomatic Complexity of 41. The configured cyclomatic complexity threshold is 10.](https://app.codacy.com/gh/Knawat/dropshipping-woocommerce/pullRequest?prid=6189509)

File: includes/class-dropshipping-woocommerce-importer.php (hunk 3 above)
Comment (HazemKhaled) [inline includes/class-dropshipping-woocommerce-importer.php:543]: Codacy found an issue: [The method set_variation_data() has an NPath complexity of 588. The configured NPath complexity threshold is 200.](https://app.codacy.com/gh/Knawat/dropshipping-woocommerce/pullRequest?prid=6189509)

File: includes/class-dropshipping-woocommerce-importer.php (hunk 11 above)
Comment (HazemKhaled) [inline includes/class-dropshipping-woocommerce-importer.php:258]: Codacy found an issue: [The method get_formatted_product() has 189 lines of code. Current threshold is set to 100. Avoid really long methods.](https://app.codacy.com/gh/Knawat/dropshipping-woocommerce/pullRequest?prid=6189509)

File: templates/admin/dropshipping-woocommerce-settings.php [hunk 13]
```diff
@@ -1,166 +1,190 @@
...
-						</p>
-					</td>
-				</tr>
+            <tr class="knawat_dropshipwc_row">
+                <th scope="row">
+					<?php _e( 'Knawat Consumer Key', 'dropshipping-woocommerce' ); ?>
+                </th>
+                <td>
+                    <input class="mp_consumer_key regular-text" name="knawat[mp_consumer_key]" type="text" value="<?php if ( $mp_consumer_key != '' ) {
+						echo $mp_consumer_key;
```
Comment (HazemKhaled) [inline templates/admin/dropshipping-woocommerce-settings.php:31]: Codacy found an issue: [Use of echo language construct is discouraged.](https://app.codacy.com/gh/Knawat/dropshipping-woocommerce/pullRequest?prid=6189509)

File: templates/admin/dropshipping-woocommerce-settings.php [hunk 14]
```diff
@@ -1,166 +1,190 @@
...
+					<?php _e( 'Categorize Products', 'dropshipping-woocommerce' ); ?>
+                </th>
+                <td>
+                    <select name="knawat[categorize_products]" required="required">
+                        <option value="no" <?php selected( 'no', $categorize_products, true ) ?>><?php esc_html_e('No'); ?></option>
+                        <option value="yes" <?php selected( 'yes', $categorize_products, true ) ?>><?php esc_html_e('Yes'); ?></option>
+                    </select>
+                    <p class="description" id="product_batch-description">
+						<?php
+						_e( 'Select yes to fetch product categories', 'dropshipping-woocommerce' );
```
Comment (HazemKhaled) [inline templates/admin/dropshipping-woocommerce-settings.php:88]: Codacy found an issue: [All output should be run through an escaping function (like esc_html_e() or esc_attr_e()), found '_e'.](https://app.codacy.com/gh/Knawat/dropshipping-woocommerce/pullRequest?prid=6189509)

File: templates/admin/dropshipping-woocommerce-settings.php [hunk 15]
```diff
@@ -1,166 +1,190 @@
...
-						</p>
-					</td>
-				</tr>
+            <tr class="knawat_dropshipwc_row">
+                <th scope="row">
+					<?php _e( 'Knawat Consumer Secret', 'dropshipping-woocommerce' ); ?>
+                </th>
+                <td>
+                    <input class="mp_consumer_secret regular-text" name="knawat[mp_consumer_secret]" type="text" value="<?php if ( $mp_consumer_secret != '' ) {
+						echo $mp_consumer_secret;
```
Comment (HazemKhaled) [inline templates/admin/dropshipping-woocommerce-settings.php:50]: Codacy found an issue: [Use of echo language construct is discouraged.](https://app.codacy.com/gh/Knawat/dropshipping-woocommerce/pullRequest?prid=6189509)

File: templates/admin/dropshipping-woocommerce-settings.php [hunk 16]
```diff
@@ -1,166 +1,190 @@
 <?php
 // If this file is called directly, abort.
-if ( ! defined( 'ABSPATH' ) ) exit;
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
```
Comment (HazemKhaled) [inline templates/admin/dropshipping-woocommerce-settings.php:4]: Codacy found an issue: [Use of exit language construct is discouraged.](https://app.codacy.com/gh/Knawat/dropshipping-woocommerce/pullRequest?prid=6189509)

File: templates/admin/dropshipping-woocommerce-settings.php (hunk 15 above)
Comment (HazemKhaled) [inline templates/admin/dropshipping-woocommerce-settings.php:50]: Codacy found an issue: [All output should be run through an escaping function (see the Security sections in the WordPress Developer Handbooks), found '$mp_consumer_secret'.](https://app.codacy.com/gh/Knawat/dropshipping-woocommerce/pullRequest?prid=6189509)

File: includes/class-dropshipping-woocommerce-common.php [hunk 17]
```diff
@@ -9,7 +9,9 @@
  * @subpackage Knawat_Dropshipping_Woocommerce/includes
  */
 // Exit if accessed directly
-if ( ! defined( 'ABSPATH' ) ) exit;
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
```
Comment (HazemKhaled) [inline includes/class-dropshipping-woocommerce-common.php:13]: Codacy found an issue: [Use of exit language construct is discouraged.](https://app.codacy.com/gh/Knawat/dropshipping-woocommerce/pullRequest?prid=6189509)

File: templates/admin/dropshipping-woocommerce-settings.php [hunk 18]
```diff
@@ -1,166 +1,190 @@
...
+						<?php
+						_e( 'Products batch size for import products from knawat.com', 'dropshipping-woocommerce' );
+						?>
+                    </p>
+                </td>
+            </tr>
+
+            <tr class="knawat_dropshipwc_row">
+                <th scope="row">
+					<?php _e( 'Categorize Products', 'dropshipping-woocommerce' ); ?>
```
Comment (HazemKhaled) [inline templates/admin/dropshipping-woocommerce-settings.php:79]: Codacy found an issue: [All output should be run through an escaping function (like esc_html_e() or esc_attr_e()), found '_e'.](https://app.codacy.com/gh/Knawat/dropshipping-woocommerce/pullRequest?prid=6189509)

File: templates/admin/dropshipping-woocommerce-settings.php (hunk 13 above)
Comment (HazemKhaled) [inline templates/admin/dropshipping-woocommerce-settings.php:31]: Codacy found an issue: [All output should be run through an escaping function (see the Security sections in the WordPress Developer Handbooks), found '$mp_consumer_key'.](https://app.codacy.com/gh/Knawat/dropshipping-woocommerce/pullRequest?prid=6189509)

**Review by hamusthekiller — CHANGES_REQUESTED** [review]
Please check the comments I left and try to modify it.
Thank You.

**Review by HazemKhaled — APPROVED** [review]

**Pull request merged**